endpointmanager: add EndpointManager type and remove package-level variables

ianvernon opened the pull request:
This is part of the effort to make Cilium more unit-testable overall. Having global state is not unit-testable by its nature. As such, refactor the fields from `pkg/endpointmanager` into a type that is passed around explicitly from `daemon` to various subsystems that need it. When packages need to use functionality from endpointmanager, it is via an interface. This allows for more unit-testable code in consumers of endpointmanager as well. 

TODO still:

- [x] address hound's comments / add additional documentation / clean up some interface names
- [ ] rebase into one commit - I split this up more for reviewing purposes, as it's quite hard to do this in an atomic way without one singular commit. Squashing into a single commit is best to ensure there are no issues while bisecting.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8742)
<!-- Reviewable:end -->

Commit:
endpointmanager: add EndpointManager type

Refactor the global fields in endpointmanager to be a type. Refactor functions
without a receiver to act upon a globally accessible instance of this type,
preserving existing behavior. No functional changes intended here.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
workloads: plumb endpointmanager as a field into each workload implementation

This removes use of the endpointmanager as a global in this package.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
cilium-health: remove use of GlobalEndpointManager

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -224,7 +224,7 @@ func CleanupEndpoint() {
 // CleanupEndpoint() must be called before calling LaunchAsEndpoint() to ensure
 // cleanup of prior cilium-health endpoint instances.
-func LaunchAsEndpoint(baseCtx context.Context, owner regeneration.Owner, n *node.Node, mtuConfig mtu.Configuration) (*Client, error) {
+func LaunchAsEndpoint(baseCtx context.Context, owner regeneration.Owner, n *node.Node, mtuConfig mtu.Configuration, epMgr *endpointmanager.EndpointManager) (*Client, error) {
```
This does not need an `endpointmanager`, it needs a interface that implements `AddEndpoint`

ianvernon replied:
@aanm agreed. Will follow-up with commits that allow us to inject the dependency.

aanm reviewed (commented):
Similar to the comment that I've given, to avoid importing the `endpointmanager` all over the place, one can use interfaces that implements the methods required to have a good functionality.

Commit:
endpointmanager: remove global endpointmanager

Now that all consumers in Cilium are using an instance of the EndpointManager
type, we can remove the need to have the global one.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
proxy: plumb endpointmanager from daemon

This removes use of the global endpointmanager directly from the proxy
package.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
datapath/maps: refactor to avoid pulling in endpointmanager dependency directly

This allows for injection of the endpointmanager dependency into said package,
as well as removes use of the global endpointmanager from the package.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
cleanup: allow ability to add cleanup functions from daemon

Only add the cleanup function to cleanup endpoint-related information when
running in tandem with Flannel once the daemon's endpointManager field is
initialized.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
daemon: add endpointManager field to Daemon type

This removes widespread use of the global endpointmanager.

Signed-off by: Ian Vernon <ian@cilium.io>

jrajahalme reviewed (approved):
LGTM

joestringer commented on daemon/daemon.go:
```diff
@@ -734,6 +734,17 @@ func NewDaemon(dp datapath.Datapath) (*Daemon, *endpointRestoreState, error) {
+
+	// Cleanup flannel on exit
+	cleanupFuncs.Add(func() {
+		if option.Config.FlannelUninstallOnExit {
```
nit: Could probably swap this and the line above to only install the run function in this case, rather than always running it.

ianvernon replied:
Fixed in latest commit!

joestringer reviewed (approved):
This is a really nice cleanup!

Commit:
daemon: only add Flannel cleanup func when running with Flannel

No functional changes.

Signed-off by: Ian Vernon <ian@cilium.io>

raybejjani commented on pkg/workloads/config.go:
```diff
@@ -97,11 +98,11 @@ type allocatorInterface interface {
 // with the options provided in opts
-func Setup(a allocatorInterface, workloadRuntimes []string, opts map[WorkloadRuntimeType]map[string]string) error {
-	return setup(a, workloadRuntimes, opts, false)
+func Setup(a allocatorInterface, workloadRuntimes []string, opts map[WorkloadRuntimeType]map[string]string, epMgr *endpointmanager.EndpointManager) error {
```
An ignorable nit. It might be slightly more consistent to pass `epMgr` as the second parameter, since we trailed configuration variables (where invoked, both `a` and `epMgr` are members on the daemon type.

ianvernon replied:
Done!

raybejjani commented on pkg/proxy/epinfo.go:
```diff
@@ -28,8 +28,14 @@ var (
 )
 
+// EndpointLookup is any type which maps from IP to the endpoint owning that IP.
+type EndpointLookup interface {
```
the DNSProxy could also use this. `Daemon.lookupEPByIP` (https://github.com/cilium/cilium/blob/master//daemon/fqdn.go#L322:18) is basically wrapping this so we could pass endpointmanager directly into the proxy in `StartDNSProxy` (https://github.com/cilium/cilium/blob/master/daemon/fqdn.go#L237)

ianvernon replied:
Created https://github.com/cilium/cilium/issues/9008

raybejjani commented on daemon/datapath.go:
```diff
@@ -335,3 +337,39 @@ func (d *Daemon) writeNetdevHeader(dir string) error {
+
+// EndpointExists returns whether the endpoint with the specified ID is managed
+// by this Cilium instance.
+func (e *EndpointMapManager) EndpointExists(endpointID uint16) bool {
```
It seems reasonable to add `EndpointExists` and `HasGlobalCT` to endpoint manager.

ianvernon replied:
We has this for `HasGlobalCT`, but not for `EndpointExists`. I'll look into seeing how I can play around with the code to make this less boilerplate-y.

ianvernon replied:
I've added `EndpointExists` to the `EndpointManager`, and now the `EndpointMapManager` is a composition of `*endpointmanager.EndpointManager`.

raybejjani commented on pkg/datapath/maps/map.go:
```diff
@@ -45,84 +42,57 @@ var (
 // This interface is provided to abstract epmanager/filesystem access for unit
 // testing.
 type endpointManager interface {
```
It seems reasonable to add `EndpointExists` and `HasGlobalCT` to endpoint manager directly, but then use this interface with those 2 functions to avoid the direct import.
`RemoveDatapathMapping` and `RemoveMapPath` seem very restricted to this package, and could just be utility functions inside it? That would also make it more clear what the thing implementing the package-local `endpointManager ` is. I might be missing something but loading the interface up seems like it only causes more boilerplate. Perhaps we have a plan to move those last two functions into endpoint manager at some point?

ianvernon replied:
Perhaps I can split up things to be two interfaces instead of just one to avoid having to wrap endpointmanager and so that I can pass in the EndpointManager directly.

ianvernon replied:
Had to look at the code again, but the reason that `RemoveDatapathMapping` are `RemoveMapPath` are part of the interface is that the unit tests have their own implementation of these which does not interact with the filesystem. So, storing them as utility functions would break the testing there (and would require having BPF imports and filesystem interactions).

raybejjani replied:
ah, I see... yeah... bummer but 🚢 . It wasn't a big deal in any case.

raybejjani reviewed (approved):
Nothing blocking, just some things that seemed off to me.

Commit:
workloads: change ordering of `Setup`

Keep configuration variables at the end of the parameter list for consistency.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
datapath: `EndpointMapManager` is a composition of `EndpointManager`

* Add the `EndpointExists` function to act on `EndpointManager`
* Have the `EndpointMapManager` be a composition of the `EndpointManager` type
so that boilerplate functions can be removed.

Signed-off by: Ian Vernon <ian@cilium.io>
